rust/src/lib.rs

```diff
@@ -827,31 +827,126 @@ type Port = u16;
+            })
     }
 }
 
 #[wasm_bindgen]
 #[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
-pub struct Ipv6(Vec<u8>);
+pub struct Ipv6([u8; 16]);
 
 to_from_bytes!(Ipv6);
 
 #[wasm_bindgen]
 impl Ipv6 {
-    pub fn new(data: Vec<u8>) -> Self {
-        Self(data)
+    pub fn new(data: Vec<u8>) -> Result<Ipv6, JsError> {
+        Self::new_impl(data).map_err(|e| JsError::from_str(&e.to_string()))
+    }
+
+    pub (crate) fn new_impl(data: Vec<u8>) -> Result<Ipv6, DeserializeError> {
+        use std::convert::TryInto;
+        data[..16]
+            .try_into()
+            .map(Self)
+            .map_err(|_e| {
+                let cbor_error = cbor_event::Error::WrongLen(16, cbor_event::Len::Len(data.len() as u64), "Ipv6 address length");
+                DeserializeError::new("Ipv6", DeserializeFailure::CBOR(cbor_error))
+            })
+    }
+}
+
+#[wasm_bindgen]
+#[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub struct URL(String);
+
+to_from_bytes!(URL);
+
+#[wasm_bindgen]
+impl URL {
+    pub fn new(url: String) -> Result<URL, JsError> {
+        Self::new_impl(url).map_err(|e| JsError::from_str(&e.to_string()))
+    }
+
+    pub (crate) fn new_impl(url: String) -> Result<URL, DeserializeError> {
+        if url.len() <= 64 {
+            Ok(Self(url))
+        } else {
+            Err(DeserializeError::new("URL", DeserializeFailure::OutOfRange{
+                min: 0,
+                max: 64,
+                found: url.len(),
+            }))
+        }
+    }
+}
+
+#[wasm_bindgen]
+#[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub struct DNSRecordAorAAAA(String);
+
+to_from_bytes!(DNSRecordAorAAAA);
+
+#[wasm_bindgen]
+impl DNSRecordAorAAAA {
+    pub fn new(dns_name: String) -> Result<DNSRecordAorAAAA, JsError> {
+        Self::new_impl(dns_name).map_err(|e| JsError::from_str(&e.to_string()))
+    }
+
+    pub (crate) fn new_impl(dns_name: String) -> Result<DNSRecordAorAAAA, DeserializeError> {
+        if dns_name.len() <= 64 {
+            Ok(Self(dns_name))
+        } else {
+            Err(DeserializeError::new("DNSRecordAorAAAA", DeserializeFailure::OutOfRange{
+                min: 0,
+                max: 64,
+                found: dns_name.len(),
+            }))
+        }
     }
 }
 
-type DnsName = String;
+#[wasm_bindgen]
+#[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub struct DNSRecordSRV(String);
+
+to_from_bytes!(DNSRecordSRV);
+
+#[wasm_bindgen]
+impl DNSRecordSRV {
+    pub fn new(dns_name: String) -> Result<DNSRecordSRV, JsError> {
+        Self::new_impl(dns_name).map_err(|e| JsError::from_str(&e.to_string()))
+    }
+
+    pub (crate) fn new_impl(dns_name: String) -> Result<DNSRecordSRV, DeserializeError> {
+        if dns_name.len() <= 64 {
+            Ok(Self(dns_name))
+        } else {
+            Err(DeserializeError::new("DNSRecordSRV", DeserializeFailure::OutOfRange{
+                min: 0,
+                max: 64,
```

Comment:
Since this `64` appears in multiple places, maybe we should make it a constant somewhere? I know a lot of pool operators complain about the 64-byte limit so I can imagine it being increased in the future.